internal/dataplane/kong_client.go

```diff
@@ -340,7 +340,18 @@ func (c *KongClient) Listeners(ctx context.Context) ([]kong.ProxyListener, []kon
 	// between reading the client(s) and setting the last applied SHA via client's
 	// SetLastConfigSHA() method. It's not ideal but it should do for now.
 	c.lock.RLock()
-	for _, cl := range c.clientsProvider.GatewayClients() {
+	gwClients := c.clientsProvider.GatewayClients()
+	c.logger.V(logging.DebugLevel).Info("Getting listeners from clients", "clients", lo.Map(gwClients, func(cl *adminapi.Client, _ int) string {
+		return cl.BaseRootURL()
+	}))
+	// If there are no ready clients yet, we should return an error.
+	// REVIEW: define a certain error type to let the gateway controller to requeue the `Gateway` with custom strategy when this happen?
```

Comment:
👍 Sounds like a good idea. And that error shouldn't be returned by the `Reconcile` func but only logged in the controller. It's not an application error but rather a transient error which will be fixed at some point.